pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java

```diff
@@ -178,36 +160,20 @@ public void initialize(ServiceConfiguration conf, PulsarResources pulsarResource
     @Override
     public CompletableFuture<Boolean> canLookupAsync(TopicName topicName, String role,
             AuthenticationDataSource authenticationData) {
-        CompletableFuture<Boolean> finalResult = new CompletableFuture<Boolean>();
-        canProduceAsync(topicName, role, authenticationData).whenComplete((produceAuthorized, ex) -> {
-            if (ex == null) {
-                if (produceAuthorized) {
-                    finalResult.complete(produceAuthorized);
-                    return;
-                }
-            } else {
-                if (log.isDebugEnabled()) {
-                    log.debug(
-                            "Topic [{}] Role [{}] exception occurred while trying to check Produce permissions. {}",
-                            topicName.toString(), role, ex.getMessage());
-                }
-            }
-            canConsumeAsync(topicName, role, authenticationData, null).whenComplete((consumeAuthorized, e)
-                    -> {
-                if (e == null) {
-                    finalResult.complete(consumeAuthorized);
-                } else {
+        return canProduceAsync(topicName, role, authenticationData)
+                .thenCompose(canProduce -> {
+                    if (canProduce) {
+                        return CompletableFuture.completedFuture(true);
+                    }
+                    return canConsumeAsync(topicName, role, authenticationData, null);
+                }).exceptionally(ex -> {
                     if (log.isDebugEnabled()) {
-                        log.debug(
-                                "Topic [{}] Role [{}] exception occurred while trying to check Consume permissions. {}",
-                                topicName.toString(), role, e.getMessage());
+                        log.debug("Topic [{}] Role [{}] exception occurred while trying to check produce/consume"
+                                + " permissions. {}", topicName.toString(), role, ex.getMessage());
 
                     }
-                    finalResult.completeExceptionally(e);
-                }
-            });
-        });
-        return finalResult;
+                    return null;
```

Comment:
Looks like this will change the original behaviors. We used to return an exceptional CompletableFuture so the caller will know the CompletableFuture is failed. If we return null in the `exceptionally` method, the caller will receive null and if there has someone use it, it will cause a NPE.

Reply:
fixed.